### PR TITLE
Unlock DFU using boot button after #2169

### PR DESCRIPTION
**Pull request by haslinghuis**
After merge #2169 DFU mode wasn't detected when using boot button on FC.

- Fixed `dfu` mode detection if using boot button on FC or `bl` cli command.
- Auto-reconnect has been taken out of the the new port detection loop to be active anytime. 
- Re-added manual selection for `dfu` port.
- Refactored several times

### CODE REVIEW

File: src/js/port_handler.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-const TIMEOUT_CHECK = 250 ; // With 250 it seems that it produces a memory leak and slowdown in some versions, reason unknown
+const TIMEOUT_CHECK = 500 ; // With 250 it seems that it produces a memory leak and slowdown in some versions, reason unknown
```
Comment (McGiverGim) [inline src/js/port_handler.js:3]: Good catch!! I don't know how I missed that in the earlier PR review 🤦‍♂️ I think the comment is very clear!!!

**Review by McGiverGim — APPROVED** [review]
I have tested it and it seems to work...

Only reconnect that continues failing is after a "save and reboot" button at some tab (for example failsafe). But I suppose that maybe this goes to another part of the code?

**Review by McGiverGim — CHANGES_REQUESTED** [review]
Thanks to @Asizon I have noticed two things:
- If I put the FC in DFU mode, it detects and selects it without problem. But if I disconnect the FC, it remains selected the DFU port that at this moment does not exists...
- It seems that sometimes, when connecting the FC in DFU mode, it goes to DFU, then to other port, and again to DFU. Maybe some selection is not ok.

**Comment by haslinghuis** [comment]
I fixed the issue disconnecting USB in `dfu` mode to show Manual selection.
The one-time switch between ports while going into `dfu` is triggered by removing the connection. But cannot pinpoint yet how to change.
Not until reconnecting (a new port is detected) it will build the full list and auto-connect if available.

**Comment by Asizon** [comment]
@haslinghuis seems that now all work as expected. I just noticed a different behavior on the `Exit DFU` button, before it goes to manual and find for devices, now it also reconect if it found devices, Im not sure if it need to reconect FC when ExIt DFU button is used or not, what do you think guys?

**Comment by McGiverGim** [comment]
@Asizon can you explain more?

I think that after "exit dfu" it must reconnect only if auto connect is enabled. If you have this option enabled, any time a FC is detected it must connect, maybe it comes from DFU, maybe it comes from a new FC plugged, is indifferent. And it seems that now is how it behaves, or at least in my case without auto connect it does not seem to connect automatically.

Have you auto connect enabled?

**Comment by Asizon** [comment]
@McGiverGim yes i have reconect enabled, but before it doesnt works like this, seems that now is the right way to go with it.

File: src/js/port_handler.js [hunk 2]
```diff
@@ -152,7 +143,14 @@ PortHandler.check_usb_devices = function (callback) {
...
                     data: {isDFU: true},
                 }));
+
+                self.portPickerElement.append($("<option/>", {
+                    value: 'manual',
+                    i18n: 'portsSelectManual',
+                    data: {isManual: true},
+                }));
                 self.portPickerElement.val('DFU').change();
+                i18n.localizePage();
```
Comment (McGiverGim) [inline src/js/port_handler.js:153]: I'm not too sure if localize all the page with any change is the way. I suppose that is for the "manual" port to be translated. Maybe is better to add directly the text in the correct language with `i18n.getMessage('portsSelectManual')`.

@haslinghuis is this `i18n.localizePage()` needed for something more?

**Comment by McGiverGim** [comment]
Looking at the tab, I see this message:
image

@mikeller this was done for some safety reason? Or simply to not connect when the user is trying to flash the FC? I think is not more a problem because it has a go to DFU in the first tab, but we need to remove the text or disable the auto-connect like it says. What do you think is the best option?

**Comment by mikeller** [comment]
@McGiverGim: I think this is done to make it possible for users to flash flight controllers after they have opened the firmware flasher tab - the assumption is made that if they select this tab then that is what they want, and using auto-connect to get them out of it is not what they want.

File: src/js/port_handler.js (hunk 2 above)
Comment (haslinghuis) [reply src/js/port_handler.js]: Had to use it otherwise it was blank, replaced with suggestion.

**Comment by haslinghuis** [comment]
My opinion is that the majority of users are not flashing a batch. You still can disable auto-connect if you want to stay in firmware flasher tab to accomplish that.

**Comment by McGiverGim** [comment]
I think the same, but I'm not too sure how other users use it.  
I suppose that it can be a problem if a user is trying to enter DFU mode to flash with the button in the FC and fails (I usually need to try several times), the Configurator will auto connect each time.
He can always disable auto connect, is very accessible, but the final decision is not mine.

**Comment by mikeller** [comment]
@haslinghuis: Not saying that users necessarily want to flash batches. But to me the behaviour that the user changes into firmware flasher, and the application then leaves it on the next change of detected flight controllers is totally unintuititve.

**Comment by haslinghuis** [comment]
Makes sense. Reverted exit `dfu` behavior.

**Review by McGiverGim — APPROVED** [review]
To me all seems ok now! Thanks!

**Comment by haslinghuis** [comment]
Locally refactored function `check_serial_devices` in 3 parts - which was relative easy - trying to solve the last code smell - but I have no SonarCloud to check against and don't want to upset the review.

**Comment by McGiverGim** [comment]
This was not necessary (this was an issue that was there before your changes but Sonar didn't associate it) but it's a welcome addition.
You can push it if you want or push it in a future PR, as you want.
I don't have any problem to review this again if you decide to push the change.

**Review by McGiverGim — APPROVED** [review]
Now the code is more clear. Good job! Sonar continues complaining but as I say is better than before.
You can play again with it or let as is, your decision 🤗

**Comment by haslinghuis** [comment]
Thank you! Enjoyed the challenge to make it more readable and functional for all platforms.

**Review by mikeller — APPROVED** [review]
Tested on MacOS, works fine. Good refactoring!

**Pull request merged**